primer/src/Primer/App.hs

```diff
@@ -824,24 +849,36 @@ applyProgAction prog = \case
                 td
           )
           type_
-  DeleteCon tdName vcName -> editModuleCross (qualifiedModule tdName) prog $ \(m, ms) -> do
-    m' <-
-      alterTypeDef
-        ( \td -> do
-            checkTypeNotInUse tdName td $ m : ms
-            traverseOf
-              #astTypeDefConstructors
-              ( \cons -> do
-                  unless
-                    (vcName `elem` map valConName cons)
-                    (throwError $ ConNotFound vcName)
-                  pure $ filter ((/= vcName) . valConName) cons
-              )
-              td
+  DeleteCon tdName vcName -> editModuleOfCrossType (Just tdName) prog $ \(m, ms) defName def -> do
+    def' <-
+      traverseOf
+        #astTypeDefConstructors
+        ( \cons -> do
+            unless
+              (vcName `elem` map valConName cons)
+              (throwError $ ConNotFound vcName)
+            pure $ filter ((/= vcName) . valConName) cons
         )
-        tdName
-        m
-    pure (m' : ms, Just $ SelectionTypeDef $ TypeDefSelection tdName Nothing)
+        def
+    ms' <-
+      ((m & over #moduleTypes (Map.insert defName $ TypeDefAST def')) : ms)
+        & traverseOf
+          (traversed % #moduleDefs % traversed % #_DefAST % #astDefExpr)
+          ( transformNamedCaseBranches
+              tdName
+              ( \_ bs ->
+                  pure
+                    $ bs
+                    & filter \case
+                      CaseBranch (PatCon vcName') _ _ | vcName' == vcName -> False
+                      _ -> True
+              )
+              <=< transformM \case
+                Con _ c _ | c == vcName -> emptyHole
+                e -> pure e
+          )
+    ms'' <- runFullTCPass (progSmartHoles prog) (progImports prog) ms'
```

Comment:
You say

> ### Deleting a constructor
>  * Replace all uses of the constructor with a hole
>  * Remove the corresponding branch from any pattern matches 
>  * Run a full typechecking pass
>     * In this instance, I haven't worked out precisely why it is necessary - I just know that without it, the "available actions accepted" property test fails quickly on some massive counterexample

This is somewhat surprising to me (though not shocking), and I thought I'd have a quick look to see if the cause is obvious. However, when I change this line to `let ms'' = ms'`, I don't see any errors running `cabal run primer-test -- -p 'available actions accepted'`. Is this typechecking pass what you meant? Could the problem have been something else what you fixed during development?

Reply:
In fact, adding a `unless (ms'' == ms') $ error "TC changed modules"` line also does not trigger

Reply:
Looking briefly at these errors, they seem to (almost?) all be only changes in typecaches, and even only filling in the typecache of the just-created holes. This is completely expected. To validate this, I ran `cabal run primer-test -- -p "available actions accepted"` in a loop with the below patch (remove the TC pass, and relax the test) and got drastically reduced numbers of errors. However, I did still see
- `cabal run primer-test -- -p "available actions accepted" --hedgehog-replay "122/200:aKdB3dEbAhCa3ChA3 Seed 210081999338504562 8735016956953542721"` about `WrongCaseBranches`.
I don't know whether this indicates a problem or not. I would like to at least understand why the procedure "replace uses with hole, delete branches" seems to give malformed `case`s on occasion.

<details>
<summary>The applied patch</summary>

```diff
diff --git a/primer/src/Primer/App.hs b/primer/src/Primer/App.hs
index a6f474ff..20e26868 100644
--- a/primer/src/Primer/App.hs
+++ b/primer/src/Primer/App.hs
@@ -877,7 +877,8 @@ applyProgAction prog = \case
                 Con _ c _ | c == vcName -> emptyHole
                 e -> pure e
           )
-    ms'' <- runFullTCPass (progSmartHoles prog) (progImports prog) ms'
+    --ms'' <- runFullTCPass (progSmartHoles prog) (progImports prog) ms'
+    let ms'' = ms'
     pure (ms'', Just $ SelectionTypeDef $ TypeDefSelection tdName Nothing)
   AddConField type_ con index new ->
     editModuleCross (qualifiedModule type_) prog $ \(m, ms) -> do
diff --git a/primer/test/Tests/Action/Available.hs b/primer/test/Tests/Action/Available.hs
index 278915a0..845c00e5 100644
--- a/primer/test/Tests/Action/Available.hs
+++ b/primer/test/Tests/Action/Available.hs
@@ -141,7 +141,7 @@ import Primer.Gen.Core.Raw (genName)
 import Primer.Gen.Core.Typed (WT, forAllT, genChar, genInt, propertyWT)
 import Primer.Log (PureLog, runPureLog)
 import Primer.Module (
-  Module (Module, moduleDefs),
+  Module (..), --(Module, moduleDefs),
   builtinModule,
   moduleDefsQualified,
   moduleName,
@@ -175,6 +175,7 @@ import Tests.Typecheck (
   runTypecheckTestMIn,
  )
 import Text.Pretty.Simple (pShowNoColor)
+import Primer.App.Utils (forgetProgTypecache)
 
 -- | Comprehensive DSL test.
 test_1 :: TestTree
@@ -502,7 +503,7 @@ tasty_available_actions_accepted = withTests 500
         (_, (Right _, a'')) -> ensureSHNormal a''
     ensureSHNormal a = case checkAppWellFormed a of
       Left err -> annotateShow err >> failure
-      Right a' -> TypeCacheAlpha a === TypeCacheAlpha a'
+      Right a' -> TypeCacheAlpha (forgetProgTypecache $ appProg a) === TypeCacheAlpha (forgetProgTypecache $ appProg a') --TypeCacheAlpha a === TypeCacheAlpha a'
 
 -- 'Raise' works when moving checkable terms into synthesisable position
 unit_raise_sh :: Assertion
```
</details>

Reply:
Ah, I understand that issue: we have `data T = C` and `foo = case (? : T) of _ -> ?` (note the wildcard). Since we don't change wildcards according to the above algorithm, we end up with `data T = <empty>`, `foo = case (? : T) of _ -> ?`. But TC complains about this fallback branch, as it expects an empty case `foo = case (? : T) of {}`!

The TC will fix this up fine, but it would be more consistent (with removing named branches) to manually remove such fallbacks.

Reply:
Thanks for looking in to this. I've added (a slightly cleaned-up version of) your patch, and edited OP accordingly.

Reply:
Ah, `null $ astTypeDefConstructors def'` isn't actually quite the right check. What if the fallback only covered one constructor, and this action deletes it? Then we'd have a meaningless wildcard. I've changed it to `map (PatCon . valConName) (astTypeDefConstructors def') == map caseBranchName bs'`, where `bs'` is the first element of the tuple.

This was actually caught by the property test when I tried to merge: `cabal run primer-test -- -p "actions accepted" --hedgehog-replay "417/1032:e2Id2J Seed 1333018207506774433 10605815445256691107"`.